ENH Calculate normed stress (Stress-1) in `manifold.MDS`

Micky774 opened the pull request:
#### Reference Issues/PRs
Resolves #10168 (stalled)
Resolves #12285 (stalled)
Resolves #13042 (stalled)
Resolves #18094 (stalled)

These PRs are (I believe) in chronological order.

#### What does this implement/fix? Explain your changes.
Implements the a normalized Stress function (Stress-1) for the `smacof` algorithm in `manifold.MDS`. Users can select to use the normalized stress value in place of the unnormalized raw stress (current default) through the `normalize` keyword.

#### Any other comments?
From my current understanding, normalized stress only guarantees true scale-invariance in the non-metric case -- as was the case when Kruscal first proposed it. I believe this is the correct implementation and faithful to the original context since non-metric MDS cares only about relative ordering of dissimilarities. 

Indeed it seems that metric MDS generally uses [stress majorization](https://en.wikipedia.org/wiki/Stress_majorization) which **does not** use normalized stress.  Consequently, I currently raise a warning when the user attempts to use `MDS` with `metric=True, normalize=True` warning that normalized stress is not supported for metric MDS and that unnormalized stress will be used instead.  

It's also worth noting that using `normalized=True` will change the scale of `eps`. I'm not sure how to approach this since I don't think warning the user every time they use `normalize=True` that they should double-check the scale of their `eps` is a good idea.

Sources: 
1. [Original non-metric mds](http://cda.psych.uiuc.edu/psychometrika_highly_cited_articles/kruskal_1964a.pdf
)
2. [Companion paper detailing numerical method](http://cda.psych.uiuc.edu/psychometrika_highly_cited_articles/kruskal_1964b.pdf
)

Follow-up PRs:
1. Change default value of `normalized` to `auto`
2. Deprecate `normalized`

Commit:
Update _mds.py

Implemented the normalized stress value from Borchmann's stalled PR: https://github.com/scikit-learn/scikit-learn/pull/10168
With these changes, passing normalize=True returns a meaningful stress value between 0-1. The current returned stress value is basically useless. normalize is set to False by default.

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 commented:
Re: concerns regarding _when_ to normalize stress (e.g. during each loop, or only at output):

From what I gather, since the normalized stress is only a scaled version of the raw stress, and furthermore is only scaled by a constant w.r.t to the runtime algorithm (since the denominator is fixed across all iterations and is dependent only on the input data), they achieve the same optima and I see no reason to scale the raw stress during fit. 

Therefore, I believe it is sufficient to keep the internal stress used for fitting as the raw stress, and only use the normalized stress for the user-facing stress value. This has the added benefit of being backwards-compatible w.r.t choice of tolerance for the algorithm.

Micky774 committed:
Move stress normalization to end of loop

Micky774 committed:
Updated tests, moved normalization to inner loop, added versionadd

Micky774 committed:
Added changelong entry

Micky774 commented:
@glemaitre @jnothman in case either of you two would be interested in reviewing

Micky774 committed:
Added warning when using normalized stress for metric MDS

Micky774 committed:
Merge branch 'main' into mds_normalized

thomasjpfan commented:
For the opening message, it's good to write "Resolves #18094 (stalled)" so that when this is merged all the linked PRs are also closed.

Micky774 commented:
> For the opening message, it's good to write "Resolves #18094 (stalled)" so that when this is merged all the linked PRs are also closed.

Updated -- thanks!

thomasjpfan commented:
The format is very specific. It has to be:

Resolves #18094

The word "PR" in the middle will not link properly

Micky774 commented:
> The format is very specific

Should be fixed now, sorry about that! I appreciate the clarification

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Fixed docstring typo

thomasjpfan reviewed (commented):
When stress is normalized, do we need to do anything with `dis`?

https://github.com/scikit-learn/scikit-learn/blob/bb5ab53f94637e6dc9341f6f8d817910ccec4cbb/sklearn/manifold/_mds.py#L129-L133

(Because we are dividing `stress` with `dis` to stop the iteration.)

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 commented:
> When stress is normalized, do we need to do anything with `dis`?
> 
> https://github.com/scikit-learn/scikit-learn/blob/bb5ab53f94637e6dc9341f6f8d817910ccec4cbb/sklearn/manifold/_mds.py#L129-L133
> 
> (Because we are dividing `stress` with `dis` to stop the iteration.)

Per a conversation over call, the current set up is fine in that `old_stress - stress / dis` is an approximation of the "normalized gradient" that Kruscal describes in his [companion paper](http://cda.psych.uiuc.edu/psychometrika_highly_cited_articles/kruskal_1964b.pdf). Users will need to tune `epsilon` differently when moving between `normalized_stress=True/False` but this is unnavoidable.

Micky774 committed:
Changed variable name from `normalized`-->`normalized_stress`

Micky774 committed:
Updated `normalized_stress` name in tests

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Update doc/whats_new/v1.1.rst

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Merge branch 'main' into mds_normalized

glemaitre reviewed (commented):
I think that we should also improve the documentation.

Since the normalization will have an impact on `eps`, we should be mentioning it in the docstring.

I am wondering if we don't want to add some information in the User Guide as well regarding the parameter in the "Nonmetric MDS" section.

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Implemented reviewer feedback

Micky774 committed:
Improved MDS User Guide

Micky774 commented:
> I am wondering if we don't want to add some information in the User Guide as well regarding the parameter in the "Nonmetric MDS" section.

Went ahead and added some content to the User Guide entry, and actually I think there was a mistake in the old entry since it was talking about preserving ordinality w.r.t _similarity_, so I changed that to _dissimilarity_.

Micky774 committed:
Merge branch 'main' into mds_normalized

glemaitre reviewed (approved):
LGTM

glemaitre committed:
Merge branch 'main' into mds_normalized

glemaitre commented:
I am merging `main` into the branch. It should solve the remaining failure.

Micky774 committed:
Merge branch 'main' into mds_normalized

glemaitre committed:
Merge branch 'main' into mds_normalized

thomasjpfan committed:
Merge branch 'main' into mds_normalized

thomasjpfan commented on sklearn/manifold/_mds.py:
```diff
@@ -152,6 +172,7 @@ def smacof(
     eps=1e-3,
     random_state=None,
     return_n_iter=False,
+    normalized_stress=False,
```
Can we extend the docstring for `eps` in `smacof` the same way you did for `MDS`?

thomasjpfan commented on sklearn/manifold/_mds.py:
```diff
@@ -99,7 +120,6 @@ def _smacof_single(
     for it in range(max_iter):
         # Compute distance and monotonic regression
         dis = euclidean_distances(X)
-
```
For git-diff purposes, can you revert this?

thomasjpfan commented on sklearn/manifold/_mds.py:
```diff
@@ -135,7 +156,6 @@ def _smacof_single(
                     print("breaking at iteration %d with stress %s" % (it, stress))
                 break
         old_stress = stress / dis
-
```
For git-diff purposes, can you revert this?

thomasjpfan reviewed (dismissed):
Minor comment about documentation, otherwise LGTM

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Addressed review comments

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Added parameter constraint

thomasjpfan commented on sklearn/manifold/_mds.py:
```diff
@@ -58,13 +59,20 @@ def _smacof_single(
 
+    normalized_stress : bool, default=False
+        Whether use and return normed stress value (Stress-1) instead of raw
+        stress calculated by default. Only supported in non-metric MDS.
```
I do not think the statement "Only supported in non-metric MDS." is not true here. The stress is normalized even if `metric=True`.

Since `smacof` itself does the validation, I am okay with not validating here. In that case, I suggest making it explicit here that the caller needs to validate:

> Only supported in non-metric MDS. The caller must ensure that if `normalized_stress=True` then `metric=False`.

thomasjpfan reviewed (commented):
I have a remaining concern on `normalized_stress` and how it automatically switches.

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Raised error and modified test

Micky774 committed:
Improved docstring

Micky774 committed:
Merge branch 'main' into mds_normalized

Micky774 committed:
Merge branch 'main' into mds_normalized

thomasjpfan reviewed (approved):
LGTM

dkobak commented:
Hi @Micky774. Currently the normalized stress is not allowed with metric MDS. I don't understand why. It works just fine, and is actually very convenient to have if one wants to compare metric MDS with non-metric MDS solutions. I suggest to remove this restriction in #30514 (that fixes some bugs in non-metric MDS implementation). What do you think?
